build.rs

```diff
@@ -43,15 +43,37 @@ fn main() {
     pkey.check_key().unwrap();
     assert_eq!(pkey.group().curve_name(), Some(Nid::X9_62_PRIME256V1));
 
-    let mut priv_key = pkey.private_key().to_vec();
-    if priv_key.len() == 33 && priv_key[0] == 0 {
-        priv_key.remove(0);
-    }
-    assert_eq!(priv_key.len(), 32);
+    // Private keys generated by OpenSSL have variable size but we only handle
+    // constant size. Serialization is done in big endian so if the size is less
+    // than 32 bytes, we need to prepend with null bytes.
+    // If the size is 33 bytes, this means the serialized BigInt is negative.
+    // Any other size is invalid.
+    let priv_key_hex = pkey.private_key().to_hex_str().unwrap();
+    let priv_key_vec = pkey.private_key().to_vec();
+    let key_len = priv_key_vec.len();
+
+    assert!(
+        key_len >= 30,
+        "Invalid private key (too small): {} ({:#?})",
+        priv_key_hex,
+        priv_key_vec,
+    );
```

Comment:
There's no reason to ban keys with length < 30 bytes. As far as I understand it, these are just keys with first bytes equal to zero - which openssl serializes as a variable length array.

Rejecting keys with many zeros at the beginning would introduce bias in the generation of private keys, which would be wrong from a cryptographic perspective.

If that's indeed openssl's process, in principle the probability to observe a key of length <= N bytes is `256^(N-32)` - so if you generate enough keys you should observe one with `N = 29` bytes for example.

Reply:
I thought we would expect a minimum entropy on the private key, that's why I put this test. But I'll remove it.

Reply:
That's not the definition of entropy. If a cryptographic algorithm specifies sampling an integer with a uniform distribution in `[1, N]`, then `1` is a valid sample and cannot be ruled out as "low-entropy".
- If one does that, then it's the distribution `[2, N]` instead - this actually *decreases* the entropy.
- `1` is not more special than any other specific value.

The same applies if instead of a single value like `1` one takes an arbitrary sub-range of integers, such as those which start with 3 (or more) zeros.

Note that the fact that an all-zeroes private key is forbidden is different (as zero behaves differently than other values in this case from a mathematical point-of-view) and this fact is specified by the cryptographic algorithm.